timestamp/tsp.go

```diff
@@ -0,0 +1,160 @@
+package timestamp
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/digitorus/pkcs7"
+	"github.com/digitorus/timestamp"
+	"github.com/testifysec/go-witness/cryptoutil"
+)
+
+type TSPTimestamper struct {
+	url                string
+	hash               crypto.Hash
+	requestCertificate bool
+}
+
+type TSPTimestamperOption func(*TSPTimestamper)
+
+func TimestampWithUrl(url string) TSPTimestamperOption {
+	return func(t *TSPTimestamper) {
+		t.url = url
+	}
+}
+
+func TimestampWithHash(h crypto.Hash) TSPTimestamperOption {
+	return func(t *TSPTimestamper) {
+		t.hash = h
+	}
+}
+
+func TimestampWithRequestCertificate(requestCertificate bool) TSPTimestamperOption {
+	return func(t *TSPTimestamper) {
+		t.requestCertificate = requestCertificate
+	}
+}
+
+func NewTimestamper(opts ...TSPTimestamperOption) TSPTimestamper {
+	t := TSPTimestamper{
+		hash:               crypto.SHA256,
+		requestCertificate: true,
```

Comment:
Have you had any issues with the server not returning certs?  I think I remember somewhere that some TSAs do not implement this

Reply:
Not currently, though I've only really tested with freetsa.org currently.

We may want to add some logic to check if there is no certificate and if so reject the timestamp perhaps